pkg/mcs/tso/server/server.go

```diff
@@ -99,16 +128,175 @@ func (s *Server) Context() context.Context {
+func (s *Server) tsoAllocatorLoop() {
+	defer logutil.LogPanic()
+	defer s.serverLoopWg.Done()
+
+	ctx, cancel := context.WithCancel(s.serverLoopCtx)
+	defer cancel()
+	s.tsoAllocatorManager.AllocatorDaemon(ctx)
+	log.Info("tso server is closed, exit allocator loop")
+}
+
+func (s *Server) primaryElectionLoop() {
+	defer logutil.LogPanic()
+	defer s.serverLoopWg.Done()
+
+	for {
+		if s.IsClosed() {
+			log.Info("server is closed, exit tso primary election loop")
+			return
+		}
+
+		primary, rev, checkAgain := s.participant.CheckLeader()
+		if checkAgain {
+			continue
+		}
+		if primary != nil {
+			// TODO: if enable-local-tso is true, check the cluster dc-location after the primary/leader is elected
+			// go s.tsoAllocatorManager.ClusterDCLocationChecker()
+
+			log.Info("start to watch the primary/leader", zap.Stringer("tso-primary", primary))
+			// WatchLeader will keep looping and never return unless the primary/leader has changed.
+			s.participant.WatchLeader(s.serverLoopCtx, primary, rev)
+			log.Info("the tso primary/leader has changed, try to re-campaign a primary/leader")
+		}
+
+		s.campaignLeader()
+	}
+}
+
+func (s *Server) campaignLeader() {
+	log.Info("start to campaign the primary/leader", zap.String("campaign-tso-primary-name", s.Name()))
+	if err := s.participant.CampaignLeader(s.cfg.LeaderLease); err != nil {
+		if err.Error() == errs.ErrEtcdTxnConflict.Error() {
+			log.Info("campaign tso primary/leader meets error due to txn conflict, another tso server may campaign successfully",
+				zap.String("campaign-tso-primary-name", s.Name()))
+		} else {
+			log.Error("campaign tso primary/leader meets error due to etcd error",
+				zap.String("campaign-tso-primary-name", s.Name()),
+				errs.ZapError(err))
+		}
+		return
+	}
+
+	// Start keepalive the leadership and enable TSO service.
+	// TSO service is strictly enabled/disabled by the leader lease for 2 reasons:
+	//   1. lease based approach is not affected by thread pause, slow runtime schedule, etc.
+	//   2. load region could be slow. Based on lease we can recover TSO service faster.
+	ctx, cancel := context.WithCancel(s.serverLoopCtx)
+	var resetLeaderOnce sync.Once
+	defer resetLeaderOnce.Do(func() {
+		cancel()
+		s.participant.ResetLeader()
+	})
+
+	// maintain the the leadership, after this, TSO can be service.
+	s.participant.KeepLeader(ctx)
+	log.Info("campaign tso primary/leader ok", zap.String("campaign-tso-primary-name", s.Name()))
+
+	allocator, err := s.tsoAllocatorManager.GetAllocator(tso.GlobalDCLocation)
+	if err != nil {
+		log.Error("failed to get the global tso allocator", errs.ZapError(err))
+		return
+	}
+	log.Info("initializing the global tso allocator")
+	if err := allocator.Initialize(0); err != nil {
+		log.Error("failed to initialize the global tso allocator", errs.ZapError(err))
+		return
+	}
+	defer func() {
+		s.tsoAllocatorManager.ResetAllocatorGroup(tso.GlobalDCLocation)
+		failpoint.Inject("updateAfterResetTSO", func() {
```

Comment:
Please remove it first since there is no related test.

Reply:
removed